modules/sdk/src/main/scala/org/tessellation/sdk/domain/cluster/programs/Joining.scala

```diff
@@ -129,14 +126,22 @@ sealed abstract class Joining[F[_]: Async: GenUUID: SecurityProvider: KryoSerial
       _ <- if (!canJoinCluster) NodeStateDoesNotAllowForJoining(nodeState).raiseError[F, Unit]
       else Applicative[F].unit
 
-      hasPeerId <- clusterStorage.hasPeerId(toPeer.id)
-      _ <- if (hasPeerId) PeerIdInUse(toPeer.id).raiseError[F, Unit] else Applicative[F].unit
-
-      hasPeerHostPort <- clusterStorage.hasPeerHostPort(toPeer.ip, toPeer.p2pPort)
-      _ <- if (hasPeerHostPort) PeerHostPortInUse(toPeer.ip, toPeer.p2pPort).raiseError[F, Unit]
-      else Applicative[F].unit
+      _ <- validateHandshakeConditions(toPeer)
     } yield ()
 
+  private def validateHandshakeConditions(toPeer: PeerToJoin): F[Unit] =
+    validateHandshakeConditions(toPeer.id, toPeer.ip, toPeer.p2pPort)
+
+  private def validateHandshakeConditions(req: RegistrationRequest): F[Unit] =
+    validateHandshakeConditions(req.id, req.ip, req.p2pPort)
+
+  private def validateHandshakeConditions(id: PeerId, ip: Host, p2pPort: Port): F[Unit] =
+    clusterStorage.hasPeerId(id).ifM(PeerIdInUse(id).raiseError[F, Unit], Applicative[F].unit).flatMap { _ =>
+      clusterStorage
+        .hasPeerHostPort(ip, p2pPort)
+        .ifM(PeerHostPortInUse(ip, p2pPort).raiseError[F, Unit], Applicative[F].unit)
```

Comment:
I think using for comprehension still could make it more readable. But it's up to you.

Reply:
I'm against using for..comp if it's not needed. Composing pipelines without syntax sugar makes me think better and upfront of data flow in the pipeline. for...comp looks more like imperative programming and can be misleading sometimes